File: OrchidCore/src/main/java/com/eden/orchid/api/tasks/OrchidTask.java
package com.eden.orchid.api.tasks;

import com.eden.orchid.api.options.Descriptive;
import com.eden.orchid.api.registration.Prioritized;
import lombok.Getter;

/**
 * A Runnable tailored for executing some task in Orchid. The 'name' of this OrchidTask is used on the command-line for
 * non-Javadoc builds to execute an alternative OrchidTask. The default OrchidTask builds the site once.
 *
 * @since v1.0.0
 * @orchidApi extensible
 */
public abstract class OrchidTask extends Prioritized implements Runnable, Descriptive {

    @Getter private final String name;
    @Getter private final TaskService.TaskType taskType;

    public OrchidTask(int priority, String name, TaskService.TaskType taskType) {
        super(priority);
        this.name = name;
        this.taskType = taskType;
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/tasks/OrchidTask.kt
package com.eden.orchid.api.tasks

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.options.Descriptive
import com.eden.orchid.api.registration.Prioritized

/**
 * A Runnable tailored for executing some task in Orchid. The 'name' of this OrchidTask is used on the command-line for
 * non-Javadoc builds to execute an alternative OrchidTask. The default OrchidTask builds the site once.
 *
 * @since v1.0.0
 * @orchidApi extensible
 */
abstract class OrchidTask(
    priority: Int,
    open val name: String,
    open val taskType: TaskService.TaskType
) : Prioritized(priority), Descriptive {

    abstract fun run(context: OrchidContext)
}
